src/resmom/linux/alps.c

```diff
@@ -6390,6 +6677,802 @@ alps_inventory(void)
+
+	if (basil_1_7_supported)
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_NODE, LOG_DEBUG, __func__,
+			"This Cray system supports the BASIL 1.7 protocol.");
+	else {
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_NODE, LOG_ERR, __func__,
+			"This Cray system does not support the BASIL 1.7 protocol.");
+		return;
+	}
+
+	/*
+	 * Allocate a buffer (requestBuffer_knl) for a new ALPS request (System Query).
+	 * A nonzero return value indicates failure; return at this point without proceeding
+	 * with System Query processing.
+	 */
+	if (new_alps_req_KNL() != 0)
+		return;
+
+	/* Create a System (BASIL 1.7) Query request to fetch KNL information. */
+	snprintf(requestBuffer_knl, UTIL_BUFFER_LEN, "<?xml version=\"1.0\"?>\n"
+		"<" BASIL_ELM_REQUEST " "
+		BASIL_ATR_PROTOCOL "=\"%s\" "
+		BASIL_ATR_METHOD "=\"" BASIL_VAL_QUERY "\" "
+		BASIL_ATR_TYPE "=\"" BASIL_VAL_SYSTEM "\"/>", basilversion_system);
+
+	/*
+	 * This flag is checked in response_start() (a callback function invoked during
+	 * alps_request() processing). Flow of control can arrive at response_start() from
+	 * either alps_system_KNL() or alps_inventory(). This flag helps make this
+	 * distinction.
+	 */
+	is_system_query = 1;
+	/* Issue XML Request. Parse XML Response. Populate appropriate basil.h structs. */
+	brp_knl = alps_request(requestBuffer_knl);
+	is_system_query = 0;
+	if (!brp_knl) {
+		/* Failure to get KNL Node information from ALPS. */
+		snprintf(log_buffer, sizeof(log_buffer), "ALPS System Query request failed.");
+		log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_NODE, LOG_NOTICE, __func__, log_buffer);
+		return;
+	}
+
+	if (basil_system_knl != NULL)
+		free(basil_system_knl);
+	basil_system_knl = strdup(alps_client_out);
+	if (basil_system_knl == NULL) {
+		snprintf(log_buffer, sizeof(log_buffer), "Failed to save System Query response.");
+		log_event(PBSEVENT_SYSTEM, PBS_EVENTCLASS_NODE, LOG_NOTICE, __func__, log_buffer);
+	}
+}
+
+/**
+ * @brief Issue an ENGINE query to determine if BASIL 1.7 is supported.
+ * 	  We are partially supporting the BASIL 1.7 protocol (for the System Query).
+ *	  If the BASIL 1.7 protocol is found in the query response, set a flag
+ *	  that will be checked in alps_system_KNL().
+ *
+ * @return void
+ */
+static void
+alps_engine_query_KNL(void)
+{
+	basil_response_t *brp_eng;
+	char *ver = NULL;
+
+	/*
+	 * Allocate a buffer (requestBuffer_knl) for a new ALPS request (Engine Query).
+	 * A nonzero return value indicates failure.
+	 */
+	if (new_alps_req_KNL() != 0)
+		return;
+	/* This is set to "1.7", since BASIL 1.7 has KNL information. */
+	snprintf(basilversion_system, sizeof(basilversion_system), BASIL_VAL_VERSION_1_7);
+	sprintf(requestBuffer_knl, "<?xml version=\"1.0\"?>\n"
+		"<" BASIL_ELM_REQUEST " "
+		BASIL_ATR_PROTOCOL "=\"%s\" "
+		BASIL_ATR_METHOD "=\"" BASIL_VAL_QUERY "\" "
+		BASIL_ATR_TYPE "=\"" BASIL_VAL_ENGINE "\"/>",
+		basilversion_system);
+
+	brp_eng = alps_request(requestBuffer_knl);
+	if (brp_eng != NULL) {
```

Comment:
can we combine these two lines to avoid redundant memory access?

Reply:
Done!